pkg/logs/decoder/line_parser.go

```diff
@@ -0,0 +1,68 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-2020 Datadog, Inc.
+
+package decoder
+
+import (
+	"github.com/DataDog/datadog-agent/pkg/logs/parser"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+// LineParser e
+type LineParser interface {
+	Handle(input *DecodedInput)
+	Start()
+	Stop()
+}
+
+// SingleLineParser makes sure that multiple lines from a same content
+// are properly put together.
+type SingleLineParser struct {
+	parser      parser.Parser
+	inputChan   chan *DecodedInput
+	lineHandler LineHandler
+}
+
+// NewSingleLineParser returns a new MultiLineHandler.
+func NewSingleLineParser(parser parser.Parser, lineHandler LineHandler) *SingleLineParser {
+	return &SingleLineParser{
+		parser:      parser,
+		inputChan:   make(chan *DecodedInput),
+		lineHandler: lineHandler,
+	}
+}
+
+// Handle puts all new lines into a channel for later processing.
+func (p *SingleLineParser) Handle(input *DecodedInput) {
+	p.inputChan <- input
+}
+
+// Start starts the parser.
+func (p *SingleLineParser) Start() {
+	p.lineHandler.Start()
+	go p.run()
+}
+
+// Stop stops the parser.
+func (p *SingleLineParser) Stop() {
+	close(p.inputChan)
+	p.lineHandler.Stop()
+}
+
+// run consumes new lines and processes them.
+func (p *SingleLineParser) run() {
+	for input := range p.inputChan {
+		p.process(input)
+	}
+}
+
+func (p *SingleLineParser) process(input *DecodedInput) {
+	// Just parse an pass to the next step
+	content, status, timestamp, _, err := p.parser.Parse(input.content)
+	if err != nil {
+		log.Debug(err)
+	}
+	p.lineHandler.Handle(NewMessage(content, status, input.rawDataLen, timestamp))
```

Comment:
I totally agree that it is likely to be a bit (I will try to find out how much) slower. Initially I thought about having some kind of fast path dropping unnecessary step (like the no-op parser) on certain condition. However in the current code base it would have either end up in something a bit hacky, or IMHO a too big change for a single PR. After that thinking step I tried to write the current PR a step to a more generic pipeline like :
` [decoder]-(*Message)-> step #1 -(*Message)-> [...] -(*Message)-> [output]`
So only mandatory block for a given source would then be instantiated, with a unique type circulating between blocks and then we won't get useless allocation for blocks that wouldn't ever be instantiated, and in that extend we should be able, for base cases (like flat file tailing, noop parser, no multiline log), to keep the same performance as we do today.
On the "new" block we could implement once that's done there have been (surprisingly) a high number of request to support utf-16 encoded log, I think if we do rework the pipeline that way, we could then have an "encoding" block usable for all log source, only enabled based on a to-be-defined config knob value.
This is really open for discussion, I was about to start an RFC on the topic with the idea of streamlining the logs processing pipeline to describe & discuss what's written above.

Reply:
Interested on this RFC!